spylib/token.py

```diff
@@ -0,0 +1,428 @@
+    ShopifyError,
+    ShopifyExceedingMaxCostError,
+    ShopifyThrottledError,
+    not_our_fault,
+)
+
+
+class AssociatedUser(BaseModel):
+    id: int
+    first_name: str
+    last_name: str
+    email: str
+    email_verified: bool
+    account_owner: bool
+    locale: str
+    collaborator: bool
+
+
+class OfflineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+
+
+class OnlineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+    expires_in: int
+    associated_user_scope: str
+    associated_user: AssociatedUser
+
+
+class Token(ABC, BaseModel):
+    """
+    Abstract class for token objects. This should never be extended, as you
+    should either be using the OfflineToken or the OnlineToken.
+    """
+
+    store_name: str
+    scope: List[str]
+    access_token: Optional[str]
+    access_token_invalid: bool
+    client: AsyncClient
+    oauth_url: str
+    api_url: str
+    api_version: str = '2021-04'
+    rest_bucket_max: int = 80
+    rest_leak_rate: int = 4
+    graphql_bucket_max: int = 1000
+    graphql_leak_rate: int = 50
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        store_name: str,
+        scope: List[str] = [],
+        access_token: Optional[str] = None,
+        api_version: str = '2021-04',
+        rest_bucket_max: int = 80,
+        rest_leak_rate: int = 4,
+        graphql_bucket_max: int = 1000,
+        graphql_leak_rate: int = 50,
+        **kwargs,
+    ) -> None:
+
+        super().__init__(
+            **{
+                'store_name': store_name,
+                'scope': scope,
+                'access_token': access_token,
+                'api_version': api_version,
+                'rest_bucket_max': rest_bucket_max,
+                'rest_leak_rate': rest_leak_rate,
+                'graphql_bucket_max': graphql_bucket_max,
+                'graphql_leak_rate': graphql_leak_rate,
+                'access_token_invalid': False,
+                'client': AsyncClient(),
+                'oauth_url': f'https://{store_name}.myshopify.com/admin/oauth/access_token',
+                'api_url': f'https://{store_name}.myshopify.com/admin/api/{api_version}',
+            }
+        )
+
+    @classmethod
+    async def new(
+        cls,
+        store_name: str,
+        client_id: str,
+        client_secret: str,
+        code: str,
+    ):
+        """
+        Creates a new instance of the object and requests the token from the API
+        as it requires an async call.
+        """
+        token = cls(store_name=store_name)
+        await token.reset_token(client_id, client_secret, code)
+        return token
+
+    @abstractmethod
+    async def save_token(self):
+        """
+        This method handles saving the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        raise NotImplementedError("Please Implement the save_token method")
```

Comment:
It's an abstract method so you don't need to do this manually. The class will do it for you if this method is not overwritten.
```suggestion
        pass